sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -1099,6 +1099,17 @@ object SQLConf {
       .intConf
       .createWithDefault(SHUFFLE_SPILL_NUM_ELEMENTS_FORCE_SPILL_THRESHOLD.defaultValue.get)
 
+  val UNION_IN_SAME_PARTITION =
+    buildConf("spark.sql.unionInSamePartition")
+      .internal()
+      .doc("When true, Union operator will union children results in the same corresponding " +
+        "partitions if they have same partitioning. This eliminates unnecessary shuffle in later " +
+        "operators like aggregation. Note that because non-deterministic functions such as " +
+        "monotonically_increasing_id are depended on partition id. By doing this, the values of " +
```

Comment:
Seems we have wanted to make sure non-deterministic functions have same values after union. Once we union children in same partitions, the values of such functions can be changed. So I added this config to control it. Default config is false.

Reply:
I'm a bit not convinced by the reason and behavior of keeping the value of non-deterministic functions after an union. Like in the following queries:

```scala
val df1 = spark.range(10).select(monotonically_increasing_id())
val df2 = spark.range(10).select(monotonically_increasing_id())
val union = df1.union(df2)
```

Now we keep the values of `monotonically_increasing_id` returned by `df1`, `df2` and `union` are the same. However, as non-deterministic functions, the values changing by data layout/sequence sounds still reasonable.

Anyway that is current behavior and this config need users to enable this feature explicitly.